src/agent/block_store_services/block_store_fs.js

```diff
@@ -58,171 +60,178 @@ class BlockStoreFs extends BlockStoreBase {
+     */
+    async _read_block(block_md) {
+        const fs_context = this.fs_context;
         const block_path = this._get_block_data_path(block_md.id);
-        const meta_path = this._get_block_meta_path(block_md.id);
-        dbg.log1('fs read block', block_path);
-        return Promise.all([
-                fs.promises.readFile(block_path),
-                fs.promises.readFile(meta_path),
-            ])
-            .then(([data_file, meta_file]) => ({
-                block_md: block_md,
-                data: data_file,
-            }))
-            .catch(err => {
-                if (err.code === 'ENOENT') {
-                    dbg.error('got error when reading', block_md, '. checking if root_path exists', err.message);
-                    this._test_root_path_exists();
+
+        try {
+            const { data, stat } = await nb_native().fs.readFile(fs_context, block_path, { read_xattr: true });
+
+            // read md from xattr
+            let block_md_from_fs = try_parse_block_md(stat.xattr[config.BLOCK_STORE_FS_XATTR_BLOCK_MD]);
+
+            // if not able to parse md from xattr, fallback to reading .meta file content which was the old model.
+            if (!block_md_from_fs) {
+                try {
+                    const meta_path = this._get_block_meta_path(block_md.id);
+                    const { data: meta_data } = await nb_native().fs.readFile(fs_context, meta_path);
+                    block_md_from_fs = try_parse_block_md(meta_data);
+                } catch (err) {
+                    // noop
                 }
-                throw err;
-            });
+            }
+
+            return { block_md: block_md_from_fs || block_md, data };
+
+        } catch (err) {
+            this._test_root_path_exists(err);
+        }
     }
 
-    _write_block(block_md, data) {
-        let overwrite_stat;
-        let md_overwrite_stat;
+    /**
+     * @param {nb.BlockMD} block_md
+     * @param {Buffer} data
+     * @param {{ ignore_usage?: boolean }} [options]
+     * @returns {Promise<void>}
+     */
+    async _write_block(block_md, data, options) {
+        const fs_context = this.fs_context;
         const block_path = this._get_block_data_path(block_md.id);
-        const meta_path = this._get_block_meta_path(block_md.id);
-        const block_md_to_store = _.pick(block_md, 'id', 'digest_type', 'digest_b64');
+        const is_test_block = Boolean(options?.ignore_usage);
+
+        /** @type {nb.NativeFSStats} */
+        let overwrite_stat;
+
+        /** @type {nb.NativeFSXattr} */
+        const xattr = {};
+
+        // fsync is needed before actually setting the migrate trigger
+        let xattr_need_fsync = false;
+
+        // set the block md xattr
+        const block_md_to_store = _.pick(block_md, 'id', 'digest_type', 'digest_b64', 'mapping_info');
         const block_md_data = JSON.stringify(block_md_to_store);
+        xattr[config.BLOCK_STORE_FS_XATTR_BLOCK_MD] = block_md_data;
```

Comment:
on overwrite, shouldn't we also delete the old meta_path file?

Reply:
I originally thought I can skip that because we do not actually expect block overwrites on our flows, and it can only occur during a retry of some sort (even then not sure it can happen). Looking back I think overwrite block could have returned an error, but as we chose to avoid erroring and just handle the case, I will fix. Thanks